enhancements/installer/aws-custom-edge-machineset-local-zones.md

```diff
@@ -0,0 +1,933 @@
+---
+title: aws-custom-edge-machineset-local-zones
+authors:
+  - "@mtulio"
+reviewers:
+  - "@rvanderp3"
+  - "@faermanj"
+  - "@patrickdillon"
+  - "@elmiko"
+  - "@JoelSpeed"
+  - "@sdodson"
+approvers:
+  - "@patrickdillon"
+  - "@JoelSpeed"
+  - "@sdodson"
+api-approvers:
+  - None
+creation-date: 2022-09-05
+last-updated: 2022-09-05
+tracking-link:
+  - https://issues.redhat.com/browse/RFE-2782
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Allow Customer-Provisioned edge machine pool for AWS Local Zones
+
+## Summary
+
+Edge adoption is increasing and providers are launching many offerings that fill
+that area. AWS launched the [Local Zones](https://aws.amazon.com/about-aws/global-infrastructure/localzones/)
+and the [Wavelength](https://aws.amazon.com/wavelength/features/) zones to extend the
+VPC network to the metropolitan area from big centers, and for Radio Access Network,
+respectively, to allow application developers to deliver low-latency applications
+at scale to the end users.
+
+As the resources in that locations were designed for specific workloads, and
+also have limited and more expensive cloud services, cluster operators should be
+able to create worker nodes on the edge networks on day zero operations.
+
+This enhancement proposes a new machine pool named `edge`, which will create
```

Comment:
Instead of asking users to set `installconfig.compute.name` to `edge`, would it make sense to validate/generate machinesets based on the `installconfig.compute.platform.aws.zones` field? Users would need to explicitly set the local zone in `installconfig.compute.platform.aws.zones` to use this functionality. It seems like `name` is more associated with the role of the node in the cluster, and would these still be considered "worker" nodes?

Reply:
Yes, it is possible. The name `edge` is used to improve the UX for the user when explicitly choosing the `edge` zones (Phase-2), and/or customizing the nodes in this pool, exclusively, without affecting the regular worker pool.

But when keeping the `edge` machine pool, I think there are other benefits of decreasing the implicit things on the installer, and the logic, when trying to fit the best compute type for this pool. In [this thread](https://github.com/openshift/enhancements/pull/1232#discussion_r979980109), @JoelSpeed and I discussed the benefits of not keeping the same compute type in `edge` pool, other than the default in regular zones.

Having said that, I can see some cons when not using `edge` pools. For example:
- if the user inputs the compute type for `installconfig.compute[name==worker].platform.aws.type` to an invalid instance type of Local Zones (which is very limited), the installer should be failed
- if the user input the compute type for `installconfig.compute[name==worker].platform.aws.type` to a valid instance type of Local Zones, the installer will deploy a more expensive cluster and will, considering current pool available in most zones, return the types used by the installer in 1+ years to use 5th-gen of instances, as the 6th-gen, default, is not generally available.
- if the user input to use `gp3` on the config, the LZ worker nodes will silently (to the installer) fail when MAPI will provision it when operating in zones that do not have this support (there is not API to query if an given zone has support for X type of EBS). If set gp2, we will use slower and more expensive block storage for regular worker nodes.
- if the user would like to deploy a subnet in the LZ with a single compute pool (worker) in Phase-2, they will need to provide the LZ zone name in `installconfig.compute[name==worker].platform.aws.zones[]`, but what will be the new default behavior of installer in this case if only LZs are defined? Currently, it will fail, otherwise, they will need to provide each regular AZs + LZs name. I think it will also impact the UX as they just want to extend to the LZ the regular IPI.

In my point of view, keeping the edge pool available to the users could provide more transparency to them in what is being deployed.

Reply:
> One thing we should keep in mind is that the idea of an edge machinepool is platform agnostic, so the idea should generally make sense across platforms. I do believe this has already been covered in this enhancement but just wanted to state it as a reminder.

Yes, it is.

Thanks for pointing it out. I am closing this discussion.